src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+// Support for HC-SR04 style ultrasonic ranging modules. See
+// https://cdn.sparkfun.com/datasheets/Sensors/Proximity/HCSR04.pdf.
+
+use std::{
+    collections::HashMap,
+    num::NonZeroU32,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc, Mutex,
+    },
+    time::Duration,
+};
+
+use crate::{
+    common::{
+        board::BoardType,
+        config::{AttributeError, ConfigType},
+        registry::{self, ComponentRegistry, Dependency},
+        sensor::{
+            GenericReadingsResult, Readings, Sensor, SensorResult, SensorT, SensorType,
+            TypedReadingsResult,
+        },
+        status::Status,
+    },
+    google, DoCommand,
+};
+
+use esp_idf_hal::{
+    delay::TickType,
+    gpio::{
+        enable_isr_service, init_isr_alloc_flags, AnyIOPin, Input, InterruptType, Output,
+        PinDriver, Pull,
+    },
+    task::notification::{Notification, Notifier},
+};
+
+use esp_idf_sys::{esp, gpio_isr_handler_add, gpio_isr_handler_remove};
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_sensor("ultrasonic", &HCSR04Sensor::from_config)
+        .is_err()
+    {
+        log::error!("HCSR04Sensor is already registered");
+    }
+}
+
+struct IsrSharedState {
+    // The state machine used to track interrupts and compute the
+    // length of the echo pulse. It holds one of the following values:
+    //
+    // 0: Starting state, ready to take a reading
+    // i64 > 0: Millisecond timestamp of first edge of echo signal
+    // -1 : Lockout state, further interrupts will be ignored
+    timestamp: AtomicI64,
+
+    // The channel the ISR will use to communicate results back to waiters.
+    notifier: Arc<Notifier>,
+}
+
+#[derive(DoCommand)]
+pub struct HCSR04Sensor {
+    // We require a board in order to ensure that ISR setup has completed
```

Comment:
This is related to the TODO down on 131. I'd like to move the ISR setup out of this file, remove what is currently being done in `pin.rs`, and start using the now `pub` interfaces from HAL to make ISR setup something that a `board` handles automatically. I'm definitely open to other approaches.

Reply:
Do we need this since the board is always configured first by default? And even if it wasn't, wouldn't it be fine just to define and register a dependency getter that requires the board?

Reply:
> Do we need this since the board is always configured first by default?

If that's the case, then no, we probably don't need it.


> And even if it wasn't, wouldn't it be fine just to define and register a dependency getter that requires the board?

I'm not entirely sure what you mean by "register a dependency getter"? Do you mean just doing `registry::get_board_from_dependencies` and then not retaining the result? I'd be fine with that.

Overall, I'd also just be fine with taking it out for now, though I expect it will creep back in later.

Reply:
to clarify, I meant using [register_dependency_getter](https://github.com/viamrobotics/micro-rdk/blob/6c3b3a5c175fe21ff2c42f258ded231305b7d1ce/src/common/registry.rs#L287)